websocket/connection.go

```diff
@@ -268,15 +268,19 @@ func (c *ManagedConnection) keepalive() error {
 
 // closeConnection closes the underlying websocket connection.
 func (c *ManagedConnection) closeConnection() error {
-	c.connectionLock.Lock()
-	defer c.connectionLock.Unlock()
-
-	if c.connection != nil {
-		err := c.connection.Close()
-		c.connection = nil
-		return err
+	// Close the connection under a read-lock to abort in-flight i/o operations and not
+	// having to wait for a read timeout to happen.
+	c.connectionLock.RLock()
+	if c.connection == nil {
+		return nil
 	}
-	return nil
+	err := c.connection.Close()
+	c.connectionLock.RUnlock()
+
+	c.connectionLock.Lock()
```

Comment:
Trying to think whether there's a potential race here: is it possible between the RUnlock and the Lock someone could reconnect? if so I think we'd null out connection without closing it

Reply:
actually maybe it's slightly worse.. right after the RUnlock the connection is closed but not nil, so we could try to use it in other methods which check if c.connection == nil under lock

Reply:
Yeah I've been thinking along similar lines. Will recheck